scale-decode-derive/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+fn unnamed_field_vals<'f>(
+    _path_to_scale_decode: &'f syn::Path,
+    fields: &'f syn::FieldsUnnamed,
+) -> (usize, impl Iterator<Item = TokenStream2> + 'f) {
+    let field_val_impls = fields.unnamed.iter().enumerate().map(|(idx, f)| {
+        let field_attrs = FieldAttrs::from_attributes(&f.attrs).unwrap_or_default();
+        let skip_field = field_attrs.skip;
+
+        // If a field is skipped, we expect it to have a Default impl to use to populate it instead.
+        if skip_field {
+            return (false, quote!(::std::default::Default::default()));
+        }
+
+        (
+            true,
+            quote!({
+                let val = vals.next().unwrap()?;
+                val.decode_as_type().map_err(|e| e.at_idx(#idx))?
+            }),
+        )
+    });
+
+    // if we skip any fields, we won't expect that field to exist in some tuple that's being given back.
+    let field_count = field_val_impls.clone().filter(|f| f.0).count();
+    let field_vals = field_val_impls.map(|v| v.1);
+
+    (field_count, field_vals)
+}
+
+fn handle_generics<'a>(
+    attrs: &TopLevelAttrs,
+    generics: &'a syn::Generics,
+) -> (syn::ImplGenerics<'a>, syn::TypeGenerics<'a>, syn::WhereClause, syn::Type) {
+    let path_to_crate = &attrs.crate_path;
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let mut where_clause = where_clause.cloned().unwrap_or(syn::parse_quote!(where));
+
+    if let Some(where_predicates) = &attrs.trait_bounds {
+        // if custom trait bounds are given, append those to the where clause.
+        where_clause.predicates.extend(where_predicates.clone());
+    } else {
+        // else, append our default bounds to each parameter to ensure that it all lines up with our generated impls and such:
+        for param in generics.type_params() {
+            let ty = &param.ident;
+            where_clause.predicates.push(syn::parse_quote!(#ty: #path_to_crate::IntoVisitor));
+            where_clause.predicates.push(syn::parse_quote!(#ty: #path_to_crate::IntoVisitor));
+            where_clause.predicates.push(syn::parse_quote!(#path_to_crate::Error: From<<<#ty as #path_to_crate::IntoVisitor>::Visitor as #path_to_crate::Visitor>::Error>));
+        }
+    }
+
+    // Construct a type to put into PhantomData<$ty>. This takes lifetimes into account too.
+    let phantomdata_type: syn::Type = {
+        let tys = generics.params.iter().filter_map::<syn::Type, _>(|p| match p {
+            syn::GenericParam::Type(ty) => {
+                let ty = &ty.ident;
+                Some(syn::parse_quote!(#ty))
+            }
+            syn::GenericParam::Lifetime(lt) => {
+                let lt = &lt.lifetime;
+                // [jsdw]: This is dumb, but for some reason `#lt ()` leads to
+                // an error (seems to output `'a, ()`) whereas this does not:
+                Some(syn::parse_quote!(::std::borrow::Cow<#lt, str>))
```

Comment:
Would it be possible that the error came from another place? I find macro's hard to reason about 🤔 

If I understand correctly, this will generate a tuple that is placed inside the `PhantomData`.
The compiler will complain about: `PhantomData<'a ()>` because there isn't a reference to attach that lifetime to. However `PhantomData<&'a ()>` should compile fine.

Maybe returning: `Some(syn::parse_quote!( & #lt ())` would do the trick here. That should output `(& 'a (),)` in the end.


I do wonder tho, do we have the `IntoVisitor` trait also implemented for references?

```rust
#[derive(scale_info::TypeInfo, codec::Encode, PartialEq, Debug, DecodeAsType)]
enum Foo<'a> {
    Wibble(&'a ()),
    Empty,
}
``` 

That seems to compile locally for `the trait `IntoVisitor` is not implemented for `&()``. 
We could, of course, tackle this as a followup, since we could also add a few examples here with `#derive[DecodeAsType]`.

Reply:
Yeahh, I spent some time with this and it was a really weird issue; the codegen seemed to spit out the correct thing, and writing it manually worked, yet for some reason I got this error in the macro that I couldn't figure out a way of avoiding!

If you can get it locally working with `&'a ()` in the codegen I'd love to see; maybe I'll have anotehr go now!

Reply:
> I do wonder tho, do we have the IntoVisitor trait also implemented for references?

I didn't bother, since decoding into references would require that the references shared a lifetime with the bytes they were coming from (or the scale info details), which I think would be fairly complicated to do (and require the DecodeAsType trait to be more tricky). So at the moment, one can do this fancy stuff with Visitors, but it's not something that's supported by the DecodeAsType interface :)